RTCEncodedFrame should have a virtual destructor

rniwa opened the pull request:
#### 5ccae9e41a6277110f2551fb0e4b012a1d274ba4
<pre>
RTCEncodedFrame should have a virtual destructor
<a href="https://bugs.webkit.org/show_bug.cgi?id=274353">https://bugs.webkit.org/show_bug.cgi?id=274353</a>

Reviewed by Chris Dumez.

Added a virtual destructor since this class has subclasses.

* Source/WebCore/Modules/mediastream/RTCEncodedFrame.h:
(WebCore::RTCEncodedFrame::~RTCEncodedFrame):

Canonical link: <a href="https://commits.webkit.org/278964@main">https://commits.webkit.org/278964@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8bdd6b6c06ff7da2cf880063b2273d14cc765688

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52093 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31421 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4509 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55366 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2805 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54397 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37810 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2513 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42390 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/1791 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54187 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29050 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44941 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23461 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26314 "Found unexpected failure with change (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2218 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/975 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48221 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2363 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56963 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27212 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2382 "Found 1 new test failure: imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/open-features-tokenization-screenx-screeny.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49787 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28448 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45062 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49017 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29353 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7615 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28189 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/mediastream/RTCEncodedFrame.h:
```diff
@@ -40,6 +40,8 @@ namespace WebCore {
 
 class RTCEncodedFrame : public RefCounted<RTCEncodedFrame> {
 public:
+    virtual ~RTCEncodedFrame() { }
```
I don't think this is necessary? The constructor is marked as protected so no-one can destroy a RTCEncodedFrame directly. Having subclasses is fine as long as their destructor gets called first, which it would since the base class constructor is not public.

cdumez replied:
Or maybe it is necessary because of the way RefCounted calls `delete`? I have to look more into it.

geoffreygaren replied:
You can't construct the base class, but you can 
* construct the subclass
* cast the subclass pointer to the base class pointer
* delete the base class pointer

In this case, I think that can mean that you fail to destroy RTCEncodedVideoFrame::m_metadata::dependencies (which is a Vector).

cdumez reviewed (approved):
Actually, looking and RefCounted::deref, I think this is indeed needed.

rniwa committed:
RTCEncodedFrame should have a virtual destructor
https://bugs.webkit.org/show_bug.cgi?id=274353

Reviewed by Chris Dumez.

Added a virtual destructor since this class has subclasses.

* Source/WebCore/Modules/mediastream/RTCEncodedFrame.h:
(WebCore::RTCEncodedFrame::~RTCEncodedFrame):

Canonical link: https://commits.webkit.org/278964@main
